audacity/audacity
# Destructive effects applied to stretched clips

Resolves: *(direct link to the issue)*

*(short description of the changes and the motivation to make the changes)*

## QA (copied from #5044)
All effects are supposed to work. Trying all of them may be too much, but a variety of types of effects should be tried (time-related effect, Nyquist, EQs and filters, noise removal, pitch shift, spectral tools, etc. Of course try them on stretched data, but also on stereo data, etc.

Note that applying an effect on a partial clip selection will create a new clip if the clip was stretched. That's wanted. The old behavior is only preserved on unstretched data.

- [x] AutoDuck
- [x] AutoDuck also works when there is stretch in the un-selected control track
- [x] Change Speed
- [x] Click Removal
- [x] EQ
- [x] Generate silence (in particular, does not create new clip boundaries at left and right edges of selection when generating over existing data and there is no stretching)
- [x] Loudness (in both modes)
- [x] Noise Reduction
- [x] Normalize
- [x] Paulstretch
- [x] Repair (Since this requires up to 128 samples only, the result will probably only be silence, and that's ok: it's a known limitation of the algorithm that rendering very tiny portions of audio only output silence)
- [x] Repeat
- [x] Reverse
- [x] Change Tempo and Sliding Scale, high and low quality
- [x] Stereo To Mono (previously reported by @dozzzzer as https://github.com/audacity/audacity/issues/5118)
- [x] Truncate Silence
- [x] Compressor
- [x] Some Nyquist effects and generators
- [x] At least one "Per Track Effect" example (any of the realtime capable, built-in or plug-in effects, but used destructively)

More for QA:
- [x] Effects applied to long enough areas, that the stretching progress indicator appears:  the indicator should fill up correctly, to 100% and then promptly vanish
- [x] Noise Reduction Analyze step, selecting stretched data and selections crossing clip boundaries with different stretch
- [x] Noise reduction analysis does not make unintended side-effects on tracks, or undo items
- [x] Test with and without the "Editing a clip can move other clips" preference (Track Behaviors)
- [x] Test all with selection of only a part of the track's duration, maybe with clip boundaries within, and different stretches
- [x] Test with silence generating a duration unequal to the selected duration
- [x] (Very minor) Click on clip boundary to merge isn't yet a hidden feature; it fails when stretch ratios differ, but makes no empty Undo item
- [x] Observe that new clip boundaries result when stretching is necessary, but if not, old behavior is preserved
- [x] Repair: stretch a mono clip by a factor of 2 and select 100 samples. You should get the error about not suitable to stretched data, not the other error that says you've selected too many samples.
- [x] #5135 is fixed by this PR
- [x] Apply a destructive effect to multiple tracks with stretch (maybe some without).  Observe just one progress bar proportioned to the number of wave tracks that have any stretch in the selection (maybe fewer than all the selected tracks).  The bar fills to completion and disappears at once, without jumping too fast over any part.
- [x] Use pencil tool _on unstretched clip(s)_ (getting this to work on stretched clips is for 3-of-6 to fix)
- [x] Effects that require sync-lock adjustment of other wave tracks that have some stretch or shrink in the selected interval
  - [x] In all of these, it should not insert splits in those tracks, and work when shortening a track, or widening a gap between clips; inserting silence into a clip is a KNOWN PROBLEM that is NOT IN THE SCOPE OF THIS PR but instead #5043 will do that
    - [x] Inserting generated silence
    - [x] Inserting generated tone, noise, chirp
    - [x] Nyquist generators, Change Speed, Paulstretch, Repeat, Change Tempo or Sliding Stretch (high and low quality)
  - [x] These two have some special case sync-lock handling and for now they WILL insert splits, though that might be reconsidered in future.
    - [x] Reverse
    - [x] Truncate Silence (Sync-locked tracks also shorten but with a cross-fade, unlike in other sync locks)




Recommended:
